GSA-TTS/FAC
# Bug: "Test Site" Banner on Production Error Pages

# Bug: "Test Site" Banner on Production Error Pages

Issue: https://github.com/GSA-TTS/FAC/issues/3567

Some context on how Django does error pages, from a similar issue we've had:
https://github.com/GSA-TTS/FAC/pull/3333

## The Problem:
1. ClamAV timeouts result in 400 errors on PDF uploads.
2. The rendered 400 page does not contain global template context, meaning:
    * The test site banner shows
    * Login information is missing
    * OMB # & expiration date are missing

## Changes:
1. Provide a useful error when ClamAV times out. This will then display to the user as a form validation error, and will (hopefully) prompt them to try again or reach out to us. 
2. Add the other standard error pages to `error_handlers.py`. This will ensure that the error pages are rendered with the context we expect. The error pages were already displaying correctly most of the time (see any intentional 404 or 403), but it's clear that certain cases will not render the same way. By having all error pages handled the same way as 500, we can avoid the edge cases.

## How to test:
First, set up your machine to display error pages without the test site banner.
1. Set your local machine to act as production. This can be done in one of the following ways:
    * Change `ENV` to "PRODUCTION" in `.env`
    * In `context_processors.py`, change `return {"ENVIRONMENT": settings.ENVIRONMENT}` to `return {"ENVIRONMENT": "PRODUCTION"}`
    * Change or add a line to `settings.py` to ensure `ENVIRONMENT` always evaluates to "PRODUCTION"
2. Turn off Django debugging. This can be done in one of the following ways:
    * Change `DJANGO_DEBUG` to "False" in `.env`
    * Change or add a line to `settings.py` to ensure `DEBUG` always evaluates to "False"

Next, verify that issues are resolved:
1. Force a ClamAV timeout on PDF uploads. One can either upload something really gross, or manually drop the timeout to something like 0.001 seconds before attempting an upload. Verify that the new validation error displays properly. 
2. Force a 400 error. I did it by raising an error in the PDF upload view, under the post. Verify that it displays _with_ the proper context (no test site banner, keeps login information). 

## Screenshots:
1. The useful error on timeouts:

![image](https://github.com/GSA-TTS/FAC/assets/91098850/6f32279a-de0a-4aff-94f1-4f17a6f88dd7)

2. The 400 page with proper context:

![image](https://github.com/GSA-TTS/FAC/assets/91098850/52e8ca5b-0898-42a0-b621-0dfaec2904e8)

## PR checklist: submitters


## PR checklist: reviewers


The larger the PR, the stricter we should be about these points.
